### PR TITLE
feat(skiplist): Add sorted skiplist builder

### PR DESCRIPTION
**Pull request by ahsanbarkati**
Add a `Builder` type in skiplist package which can be used to insert
sorted keys efficiently.

---
This change is [Reviewable](https://reviewable.io/reviews/dgraph-io/badger/1693)

### CODE REVIEW

File: skl/skl_test.go [hunk 1]
```diff
@@ -457,6 +458,26 @@ func randomKey(rng *rand.Rand) []byte {
 	return y.KeyWithTs(b, 0)
 }
 
+func TestBuilder(t *testing.T) {
+	N := 1 << 16
+	b := NewBuilder(32 << 20)
+	for i := 0; i < N; i++ {
+		buf := make([]byte, 8)
```
Comment (jarifibrahim) [inline skl/skl_test.go:465]: Move this out of the for loop.

File: skl/skl_test.go [hunk 2]
```diff
@@ -457,6 +458,26 @@ func randomKey(rng *rand.Rand) []byte {
...
+		binary.BigEndian.PutUint64(buf, uint64(i))
+		key := y.KeyWithTs(buf, 0)
+		b.Add(key, y.ValueStruct{Value: []byte("00072")})
+	}
+	l := b.s
+	for i := 0; i < N; i++ {
+		buf := make([]byte, 8)
```
Comment (jarifibrahim) [inline skl/skl_test.go:472]: This one also.

File: skl/skl_test.go [hunk 3]
```diff
@@ -457,6 +458,26 @@ func randomKey(rng *rand.Rand) []byte {
...
+		binary.BigEndian.PutUint64(buf, uint64(i))
+		key := y.KeyWithTs(buf, 0)
+		v := l.Get(key)
+		require.True(t, v.Value != nil)
```
Comment (jarifibrahim) [inline skl/skl_test.go:476]: Require.NotNil(...)

**Review by jarifibrahim — APPROVED** [review]

**Review by manishrjain — APPROVED** [review]
:lgtm:

> Reviewed 3 of 3 files at r2.
> *[Reviewable](https://reviewable.io/reviews/dgraph-io/badger/1693#-:-MYphu790Pgg9ekiAxYX:bnfp4nl)* status: all files reviewed, 5 unresolved discussions (waiting on @ahsanbarkati)

___
*[skl/skl.go, line 550 at r2](https://reviewable.io/reviews/dgraph-io/badger/1693#-MYph3RZCXWAx_L5nYnk:-MYph3RZCXWAx_L5nYnl:b-q2tomx) ([raw file](https://github.com/dgraph-io/badger/blob/84fe69b2f3a559aabd3bbae7c562eaba80e6b809/skl/skl.go#L550)):*
> ```Go
> 	if debug {
> 		if len(b.prevKey) > 0 && y.CompareKeys(k, b.prevKey) <= 0 {
> 			panic(fmt.Sprintf("new key: %s <= prev key: %s\n", y.ParseKey(k), y.ParseKey(b.prevKey)))
> ```

overflow
___
*[skl/skl.go, line 563 at r2](https://reviewable.io/reviews/dgraph-io/badger/1693#-MYphs5I1u0NFa9g5Py_:-MYphs5I1u0NFa9g5Pya:bv8akfl) ([raw file](https://github.com/dgraph-io/badger/blob/84fe69b2f3a559aabd3bbae7c562eaba80e6b809/skl/skl.go#L563)):*
> ```Go
> 	for i := 0; i < height; i++ {
> 		node := b.prev[i]
> 		node.tower[i] = s.arena.getNodeOffset(x)
> ```

can move this outside.

**Pull request merged**